KEYCLOAK-13745 add config for loading custom IdMapper class

mkausl opened the pull request.

mkausl committed:
KEYCLOAK-13745 add config for loading custom IdMapper class

mkausl committed:
fix test

mposolda commented:
@mkausl Thanks for the PR. Few points to that:

- Better test will be required. Please remove the "mock" simulation testing as we try to avoid mock-like testing. It will be good to have integration test in our arquillian testsuite. For some hints, see classes like DemoFilterServletAdapterTest, UseServletFilter (maybe some new parameter will need to be added here) and DeploymentArchiveProcessor (where the configuration is processed etc). It will be fine to create possibly subclass of InMemorySessionIdMapper and "decorate" it somehow and then in the test somehow verify that "decorated" version was used instead of the default  InMemorySessionIdMapper

- The documentation will be required. There is the section here https://www.keycloak.org/docs/latest/securing_apps/index.html#_servlet_filter_adapter, which will be nice to improve. It will be good to document all the parameters of KeycloakOIDCFilter when you're at it as there are few more, which are available and undocumented (eg. keycloak.config.file, keycloak.config.path etc). For the documentation, you can send PR to https://github.com/keycloak/keycloak-documentation/

mposolda reviewed (changes requested):
Also just a note that you can create subclass of KeycloakOIDCFilter and override the idMapper here.

mkausl committed:
remove mock tests and add integration tests

mkausl commented:
> @mkausl Thanks for the PR. Few points to that:
> 
> * Better test will be required. Please remove the "mock" simulation testing as we try to avoid mock-like testing. It will be good to have integration test in our arquillian testsuite. For some hints, see classes like DemoFilterServletAdapterTest, UseServletFilter (maybe some new parameter will need to be added here) and DeploymentArchiveProcessor (where the configuration is processed etc). It will be fine to create possibly subclass of InMemorySessionIdMapper and "decorate" it somehow and then in the test somehow verify that "decorated" version was used instead of the default  InMemorySessionIdMapper
> * The documentation will be required. There is the section here https://www.keycloak.org/docs/latest/securing_apps/index.html#_servlet_filter_adapter, which will be nice to improve. It will be good to document all the parameters of KeycloakOIDCFilter when you're at it as there are few more, which are available and undocumented (eg. keycloak.config.file, keycloak.config.path etc). For the documentation, you can send PR to https://github.com/keycloak/keycloak-documentation/

I removed the mock and added a copy of DemoFilterServletAdapterTest with the new config option. I was able to verify that the configured class was loaded, however I'm not sure how i can test this. maybe you can give me a hint?

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/DemoFilterServletAdapterWithIdMapperConfigTest.java:
```diff
@@ -0,0 +1,98 @@
+@AppServerContainer(ContainerConstants.APP_SERVER_EAP71)
+@UseServletFilter(filterName = "oidc-filter", filterClass = "org.keycloak.adapters.servlet.KeycloakOIDCFilter",
+        filterDependency = "org.keycloak:keycloak-servlet-filter-adapter", skipPattern = "/error.html", configIdMapperClass = "org.keycloak.helpers.TestInMemorySessionIdMapper")
+public class DemoFilterServletAdapterWithIdMapperConfigTest extends DemoServletsAdapterTest {
```
Instead of extending from DemoServletsAdapterTest, it will be better to either:
- extend from AbstractServletsAdapterTest and add only single test method
- change the existing DemoFilterServletAdapterTest and change the configuration of it to your idMapper

Adding completely new class, which extends from DemoServletsAdapterTest adds around 30-50 additional running tests to the testsuite just due this relatively small feature.

mposolda reviewed (changes requested):
Thanks for adding the test and sorry for long delay.

I've added one inline comment. Regarding the test, maybe to validate that your IdMapper is called, you can add some singleton class, which will extend from the default IdMapper, but in addition, it will contain some very simple state (EG. some key-value map), which will track that methods from your IdMapper were called. Then there may be addition REST endpoint, which will allow to read that state and hence verify that your IdMapper was called.

For the inspiration, you can take a look for example at the AdapterActionsFilter and possibly add some new functionality to it, which will allow to read the state from that singleton.

I know adding this test is non-trivial, but IMO it is needed to avoid having another untested functionality, which will break anytime in the future. Good luck with the test!

mposolda commented:
@mkausl Any update regarding this PR?

mkausl commented:
sorry, i did not have time to work on it for now

mposolda commented:
@mkausl np, Thanks for the update!

mposolda commented:
@mkausl I am closing this as we need to reduce the pull requests queue a bit ;) Feel free to re-open once you add the changes and documentation
